intermediate_source/pinmem_nonblock.py

```diff
@@ -0,0 +1,625 @@
+# can live anywhere in the virtual memory (both in RAM or on disk). As mentioned earlier, this has the advantage that
+# the memory seems larger than what the main memory actually is.
+#
+# Typically, when a program accesses a page that is not in RAM, a "page fault" occurs and the operating system (OS) then brings
+# back this page into RAM (_swap in_ or _page in_).
+# In turn, the OS may have to _swap out_ (or _page out_) another page to make room for the new page.
+#
+# In contrast to pageable memory, a _pinned_ (or _page-locked_ or _non-pageable_) memory is a type of memory that cannot
+# be swapped out to disk.
+# It allows for faster and more predictable access times, but has the downside that it is more limited than the
+# pageable memory (aka the main memory).
+#
+# .. figure:: /_static/img/pinmem.png
+#    :alt:
+#
+# CUDA and (non-)pageable memory
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+#   .. _pinmem_cuda_pageable_mem:
+#
+# To understand how CUDA copies a tensor from CPU to CUDA, let's consider the two scenarios above:
+# - If the memory is page-locked, the device can access the memory directly in the main memory. The memory addresses are well
+#   defined and functions that need to read these data can be significantly accelerated.
+# - If the memory is pageable, all the pages will have to be brought to the main memory before being sent to the GPU.
+#   This operation may take time and is less predictable than when executed on page-locked tensors.
+#
+# More precisely, when CUDA sends pageable data from CPU to GPU, it must first create a page-locked copy of that data
+# before making the transfer.
+#
+# Asynchronous vs. Synchronous Operations with ``non_blocking=True`` (CUDA ``cudaMemcpyAsync``)
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+#   .. _pinmem_async_sync:
+#
+# When executing a copy from a host (e.g., CPU) to a device (e.g., GPU), the CUDA toolkit offers modalities to do these
+# operations synchronously or asynchronously with respect to the host.
+#
+# In practice, when calling :meth:`~torch.Tensor.to`, PyTorch always makes a call to
+# `cudaMemcpyAsync <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79>`_.
+# If ``non_blocking=False`` (default), a ``cudaStreamSynchronize`` will be called after each and every ``cudaMemcpyAsync``, making
+# the call to :meth:`~torch.Tensor.to` blocking in the main thread.
+# If ``non_blocking=True``, no synchronization is triggered, and the main thread on the host is not blocked.
+# Therefore, from the host perspective, multiple tensors can be sent to the device simultaneously,
+# as the thread does not need to wait for one transfer to be completed to initiate the other.
+#
+# .. note:: In general, the transfer is blocking on the device side (even if it isn't on the host side):
```

Comment:
n00b qn: tbh this is something that I have always been confused about, but I can't seem to get overlap for copies.

I tried this with the following script on A100 devgpu but I can't seem to get overlap, do you know what might be wrong here?
<img width="1500" alt="Screenshot 2024-07-26 at 12 37 45 PM" src="https://github.com/user-attachments/assets/bc034d3d-513a-4ee6-bb45-b9de1bfc1d10">



<details closed>
<summary>Click for Script</summary>
<br>

```python

import torch
from torch.cuda import Stream


def benchmark_with_profiler(
    benchmark_fn,
    *benchmark_fn_args,
    **benchmark_fn_kwargs,
) -> None:
    torch._C._profiler._set_cuda_sync_enabled_val(True)
    wait, warmup, active = 1, 1, 2
    num_steps = wait + warmup + active
    rank = 0
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        schedule=torch.profiler.schedule(
            wait=wait, warmup=warmup, active=active, repeat=1, skip_first=1
        ),
        on_trace_ready=torch.profiler.tensorboard_trace_handler("./")
        if not rank
        else None,
        record_shapes=True,
        profile_memory=True,
        with_stack=False,
        with_flops=True,
        with_modules=False,
    ) as prof:
        for step_idx in range(1, num_steps + 1):
            benchmark_fn(*benchmark_fn_args, **benchmark_fn_kwargs)
            if rank is None or rank == 0:
                prof.step()

s = Stream()

torch.manual_seed(42)
t1_cpu = torch.randn(1024 ** 2 * 5, pin_memory=True)
t2_cpu = torch.randn(1024 ** 2 * 5, pin_memory=True)

assert torch.cuda.is_available()
device = torch.device("cuda", torch.cuda.current_device())

def inner():
    with torch.cuda.stream(s):
        t2_cuda = t2_cpu.to(device, non_blocking=True)
        t2_h2d_event = s.record_event()
    t1_cuda = t1_cpu.to(device, non_blocking=True)
    t1_h2d_event = torch.cuda.current_stream().record_event()
    t1_h2d_event.synchronize()
    t2_h2d_event.synchronize()

benchmark_with_profiler(inner)

```

</details>

Reply:
Seems like it's less about concurrent cpu->device copies and more about copy while doing another op.


<details closed>
<summary>Click for Script</summary>
<br>

```python
import torch
from torch.cuda import Stream


def benchmark_with_profiler(
    benchmark_fn,
    *benchmark_fn_args,
    **benchmark_fn_kwargs,
) -> None:
    torch._C._profiler._set_cuda_sync_enabled_val(True)
    wait, warmup, active = 1, 1, 2
    num_steps = wait + warmup + active
    rank = 0
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        schedule=torch.profiler.schedule(
            wait=wait, warmup=warmup, active=active, repeat=1, skip_first=1
        ),
        on_trace_ready=torch.profiler.tensorboard_trace_handler("./")
        if not rank
        else None,
        record_shapes=True,
        profile_memory=True,
        with_stack=False,
        with_flops=True,
        with_modules=False,
    ) as prof:
        for step_idx in range(1, num_steps + 1):
            benchmark_fn(*benchmark_fn_args, **benchmark_fn_kwargs)
            if rank is None or rank == 0:
                prof.step()

s = Stream()

torch.manual_seed(42)
t1_cpu_pinned = torch.randn(1024 ** 2 * 5, pin_memory=True)
t2_cpu_pageable = torch.randn(1024 ** 2 * 5, pin_memory=False)
t3_cuda = torch.randn(1024 ** 2 * 5, device="cuda:0")

assert torch.cuda.is_available()
device = torch.device("cuda", torch.cuda.current_device())

def inner():
    with torch.cuda.stream(s): # Comment to get main stream
        t2_cuda = t2_cpu_pageable.to(device, non_blocking=True)  # comment and uncomment following for pinned
        # t1_cuda = t1_cpu_pinned.to(device, non_blocking=True)
        t2_h2d_event = s.record_event()
    t3_cuda_mul = t3_cuda * t3_cuda * t3_cuda
    t1_h2d_event = torch.cuda.current_stream().record_event()
    t1_h2d_event.synchronize()
    t2_h2d_event.synchronize()

benchmark_with_profiler(inner)
```
</details>

Playing with the following code (commenting/uncommenting things and executing in same stream or different):

1. non_blocking copy from pinned tensor on dedicated stream, followed by a multiplication on main stream
![image](https://github.com/user-attachments/assets/21f438ed-ce72-4400-b214-7d82dd1510da)
2. non_blocking copy from pinned tensor on main stream, followed by a multiplication on main stream
![image](https://github.com/user-attachments/assets/f0b506f6-0b9a-43e1-92e1-d3cfd635041c)
4. non_blocking copy from pageable tensor on dedicated stream, followed by a multiplication on main stream
![image](https://github.com/user-attachments/assets/011b0f62-672d-4846-b7e1-44e138906516)

I will edit the doc to say this:

If the tensor is in pinned mem, the device has a free DMA and the copy is done in a separate stream, CUDA can execute another op in the main stream concurrently to the copy - but not another copy.

Agreed? (maybe it could do another copy on a third stream??)

Reply:
Yea I think the updated wording mostly makes sense except the last bit! Not sure whether we want to very conclusively say that another copy is not possible as I am not 100% sure whether there is absolutely no scenario under which it could do another copy as well 😅  (perhaps H2D and D2H could be overlapped??)